### PR TITLE
Connections print diagnostics via operator<<

### PR DESCRIPTION
**Pull request by ctrl-z-9000-times**
For issue #355.

I tested this with the MNIST dataset.  Here is the output:
```
$ time ./bin/mnist_sp
------------CPP SpatialPooler Parameters ------------------
iterationNum                = 0
iterationLearnNum           = 0
numInputs                   = 784
numColumns                  = 10000
numActiveColumnsPerInhArea  = -1
potentialPct                = 0.5
globalInhibition            = 1
localAreaDensity            = 0.015
stimulusThreshold           = 6
synPermActiveInc            = 0.01
synPermInactiveDec          = 0.005
synPermConnected            = 0.4
minPctOverlapDutyCycles     = 0.001
dutyCyclePeriod             = 1402
boostStrength               = 2.5
spVerbosity                 = 1
wrapAround                  = 0
version                     = 2
CPP SP seed                 = 93
Training for 60000 cycles ...
........................................................................................................................
epoch ended
inputStats SDR( 784 )
    Sparsity Min/Mean/Std/Max 0.0433673 / 0.201973 / 0.0381583 / 0.447704
    Activation Frequency Min/Mean/Std/Max 0 / 0.201971 / 0.293482 / 0.871855
    Entropy 0.476691
    Overlap Min/Mean/Std/Max 0.0496894 / 0.581518 / 0.261094 / 1

columnStats SDR( 10000 )
    Sparsity Min/Mean/Std/Max 0.015 / 0.0150001 / 7.91346e-08 / 0.015
    Activation Frequency Min/Mean/Std/Max 0 / 0.0149995 / 0.0297155 / 0.17009
    Entropy 0.722389
    Overlap Min/Mean/Std/Max 0 / 0.0194646 / 0.0706917 / 0.92

SpatialPooler Connections:
    Inputs (784) ~> Outputs (10000) via Segments (10000)
    Potential Synapses on Segment Min/Mean/Max 392 / 392 / 392
    Connected Synapses on Segment Min/Mean/Max 39 / 161.406 / 223
    Synapses Dead (0.233773%) Saturated (0%)

Testing for 10000 cycles ...
..........
Score: 91.14% 

real	1m53.984s
user	1m53.770s
sys	0m0.132s
```

### CODE REVIEW

File: src/examples/mnist/MNIST_SP.cpp [hunk 1]
```diff
@@ -58,14 +58,15 @@ class MNIST {
 
 void setup() {
 
-  input.initialize({28, 28});
+  input.initialize({28 * 28});
```
Comment (breznak) [inline src/examples/mnist/MNIST_SP.cpp]: why the change from 2D to flat? (28*28)

Comment (ctrl-z-9000-times) [reply src/examples/mnist/MNIST_SP.cpp]: It's easier to work with.  Besides, there is no topology in this example because it uses global inhibition and the radius is larger than the inputs.

**Comment by ctrl-z-9000-times** [comment]
Thank you Breznak for reviewing this!
